chatbots/shopping_buddy.py

```diff
@@ -130,6 +137,48 @@ def recommend(state: State):
     return state
 
 
+def related_router(state: State) -> str:
+    if state["recommended_product_data"] is None:
+        logger.debug("ROUTER: to the end")
+        return END
+    else:
+        logger.debug("ROUTER: find_related_products")
+        return "find_related_products"
+    
+
+def find_related_products(state: State) -> State:
+    logger.debug("----------find_related_products----------")
+    system_messages = get_related_product_preference(state["customer_preference"])
+    response = llm_with_product_tools.invoke(system_messages)
+    state["messages"] = add_messages(state["messages"], [response])
+    state["related_product_preference"] = parse_related_product_preference(response.tool_calls[0]["args"])
+    state["messages"] = add_messages(state["messages"], [
+        ToolMessage(
+            content="Related Product Preference gathered",
+            tool_call_id=state["messages"][-1].tool_calls[0]["id"],
+        )
+    ])
+    query_text = format_related_product_preference(state["related_product_preference"])
+    # vector search
+    search_result = vector_search_product(query_text, columns=["product_id", "title", "text"], num_results=3)
+    product_ids, similarity = process_search_result(search_result)
```

Comment:
loop over `state["related_product_preference"]` for searching
make sure to change format_related_product_preference so that it takes in 1 product category